%% BioMed_Central_Tex_Template_v1.06
%%                                      %
%  bmc_article.tex            ver: 1.06 %
%                                       %

%%IMPORTANT: do not delete the first line of this template
%%It must be present to enable the BMC Submission system to
%%recognise this template!!

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                     %%
%%  LaTeX template for BioMed Central  %%
%%     journal article submissions     %%
%%                                     %%
%%          <8 June 2012>              %%
%%                                     %%
%%                                     %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                                                 %%
%% For instructions on how to fill out this Tex template           %%
%% document please refer to Readme.html and the instructions for   %%
%% authors page on the biomed central website                      %%
%% https://www.biomedcentral.com/getpublished                      %%
%%                                                                 %%
%% Please do not use \input{...} to include other tex files.       %%
%% Submit your LaTeX manuscript as one .tex document.              %%
%%                                                                 %%
%% All additional figures and files should be attached             %%
%% separately and not embedded in the \TeX\ document itself.       %%
%%                                                                 %%
%% BioMed Central currently use the MikTex distribution of         %%
%% TeX for Windows) of TeX and LaTeX.  This is available from      %%
%% https://miktex.org/                                             %%
%%                                                                 %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%% additional documentclass options:
%  [doublespacing]
%  [linenumbers]   - put the line numbers on margins

%%% loading packages, author definitions

%\documentclass[twocolumn]{bmcart}% uncomment this for twocolumn layout and comment line below
\documentclass{bmcart}

%%% Load packages
\usepackage{amsthm,amsmath}
%\RequirePackage[numbers]{natbib}
%\RequirePackage[authoryear]{natbib}% uncomment this for author-year bibliography
\RequirePackage{hyperref}
\usepackage[utf8]{inputenc} %unicode support
%\usepackage[applemac]{inputenc} %applemac support if unicode package fails
%\usepackage[latin1]{inputenc} %UNIX support if unicode package fails


%% SUBMITTER PACKAGES
%% To my knowledge, the below packages were used to render lines, symbols and adequate spacing in tables.
\usepackage{booktabs} % de-congest whitespace around tables using \midrule NOT \hrule
\usepackage{siunitx}
\usepackage[strings]{underscore} % knit without DOI string errors - must not use underscores in refs or bib ref
\usepackage{fancybox} %% supp info box 
\usepackage{array,multirow}
\usepackage{scrextend} % indent availability and requirements
\usepackage{enumitem} % centered list
\usepackage{pifont} % xmark, cmark in table1
\usepackage[normalem]{ulem} %% reviewer comments (strikethrough deleted text)
\newcommand{\cmark}{\ding{51}}%
\newcommand{\xmark}{\ding{55}}%

\newcommand{\beginsupplement}{%
        \setcounter{table}{0}
        \renewcommand{\thetable}{S\arabic{table}}%
        \setcounter{figure}{0}
        \renewcommand{\thefigure}{S\arabic{figure}}%
     }


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                             %%
%%  If you wish to display your graphics for   %%
%%  your own use using includegraphic or       %%
%%  includegraphics, then comment out the      %%
%%  following two lines of code.               %%
%%  NB: These line *must* be included when     %%
%%  submitting to BMC.                         %%
%%  All figure files must be submitted as      %%
%%  separate graphics through the BMC          %%
%%  submission process, not included in the    %%
%%  submitted article.                         %%
%%                                             %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%% I have commented \def lines and included usepackage graphicx to render pictures in PDF file. 
%% undo this for submission and provide TIFF files

%\def\includegraphic{}
%\def\includegraphics{}
\usepackage{graphicx}

%%% Put your definitions there:
\startlocaldefs
\endlocaldefs

%%% Begin ...
\begin{document}

%%% Start of article front matter
\begin{frontmatter}

\begin{fmbox}
\dochead{Software}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the title of your article here     %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\title{nf-core/circrna: a portable workflow for the quantification, miRNA target prediction and differential expression analysis of circular RNAs}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the authors here                   %%
%%                                          %%
%% Specify information, if available,       %%
%% in the form:                             %%
%%   <key>={<id1>,<id2>}                    %%
%%   <key>=                                 %%
%% Comment or delete the keys which are     %%
%% not used. Repeat \author command as much %%
%% as required.                             %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\author[
  addressref={aff1},                   % id's of addresses, e.g. 
    %corref={aff1}, %% NOTE TO EDITOR: I had to comment this line to avoid bmc@corref@authorthanks undefined errors - very confusing.. 
    email={barry.digby@nuigalway.ie}  % email address
]{\inits{B.D.}\fnm{Barry} \snm{Digby}}
\author[
  addressref={aff2},
  email={Stephen.Finn@tcd.ie}
]{\inits{S.P.F}\fnm{Stephen P.} \snm{Finn}}
\author[
  addressref={aff1},
  email={pilib.obroin@nuigalway.ie}
]{\inits{P.O.B.}\fnm{Pilib} \snm{\'O Broin}}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the authors' addresses here        %%
%%                                          %%
%% Repeat \address commands as much as      %%
%% required.                                %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\address[id=aff1]{%                           % unique id
  \orgdiv{School of Mathematical and Statistical Sciences},             % department, if any
  \orgname{National University of Ireland},          % university, etc
  \city{Galway},                              % city
  \cny{Ireland}                                    % country
}
\address[id=aff2]{%
  \orgdiv{Department of Histopathology and Morbid Anatomy},
  \orgname{Trinity Translational Medicine Institute},
  %\street{},
  %\postcode{}
  \city{Dublin},
  \cny{Ireland}
}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter short notes here                   %%
%%                                          %%
%% Short notes will be after addresses      %%
%% on first page.                           %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%\begin{artnotes}
%%\note{Sample of title note}     % note to the article
%\note[id=n1]{Equal contributor} % note, connected to author
%\end{artnotes}

\end{fmbox}% comment this for two column layout

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                           %%
%% The Abstract begins here                  %%
%%                                           %%
%% Please refer to the Instructions for      %%
%% authors on https://www.biomedcentral.com/ %%
%% and include the section headings          %%
%% accordingly for your article type.        %%
%%                                           %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{abstractbox}

\begin{abstract} % abstract
\parttitle{Background} %if any
Circular RNAs (circRNAs) are a class of covalenty closed non-coding RNAs (ncRNAs) that have garnered increased attention from the research community due to their stability, tissue-specific expression and role as transcriptional modulators via sequestration of miRNAs. Currently, multiple quantification tools capable of detecting circRNAs exist, yet none delineate circRNA-miRNA interactions, and only one employs differential expression analysis. Efforts have been made to bridge this gap by way of circRNA workflows, however these workflows are limited by both the types of analyses available and computational skills required to run them.

\parttitle{Results} %if any
We present nf-core/circrna, a multi-functional, automated high-throughput pipeline implemented in nextflow that allows users to characterise the role of circRNAs in RNA Sequencing datasets via three analysis modules: (i) circRNA quantification, robust filtering and annotation (ii) miRNA target prediction of the mature spliced sequence and (iii) differential expression analysis. nf-core/circrna has been developed within the nf-core framework, ensuring robust portability across computing environments via containerisation, parallel deployment on cluster/cloud-based infrastructures, comprehensive documentation and maintenance support.

\parttitle{Conclusion}
nf-core/circrna reduces the barrier to entry for researchers by providing an easy-to-use, platform-independent and scalable workflow for circRNA analyses. Source code, documentation and installation instructions are freely available at \url{https://nf-co.re/circrna} and \url{https://github.com/nf-core/circrna}.

\end{abstract}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The keywords begin here                  %%
%%                                          %%
%% Put each keyword in separate \kwd{}.     %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{keyword}
\kwd{circular RNA}
\kwd{ceRNA}
\kwd{nextflow}
\kwd{nf-core}
\kwd{portable analysis workflow}
\end{keyword}

% MSC classifications codes, if any
%\begin{keyword}[class=AMS]
%\kwd[Primary ]{}
%\kwd{}
%\kwd[; secondary ]{}
%\end{keyword}

\end{abstractbox}
%
%\end{fmbox}% uncomment this for two column layout

\end{frontmatter}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                            %%
%% The Main Body begins here                  %%
%%                                            %%
%% Please refer to the instructions for       %%
%% authors on:                                %%
%% https://www.biomedcentral.com/getpublished %%
%% and include the section headings           %%
%% accordingly for your article type.         %%
%%                                            %%
%% See the Results and Discussion section     %%
%% for details on how to create sub-sections  %%
%%                                            %%
%% use \cite{...} to cite references          %%
%%  \cite{koon} and                           %%
%%  \cite{oreg,khar,zvai,xjon,schn,pond}      %%
%%                                            %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%%%%%%%%%%%%%%%%%%% start of article main body
% <put your article body there>

%%%%%%%%%%%%%%%%
%% Background %%
%%
\section*{Background}
CircRNAs are a class of non-coding RNAs (ncRNAs) formed by the back-splicing of precursor messenger RNA (pre-mRNA) to create a single-stranded covalently closed RNA loop structure. Initially discovered in plant viroids \cite{Sanger1976}, yeast mitochondrial RNAs \cite{ARNBERG1980}, and the hepatitis $\delta$ virus \cite{Kos1986}, circRNAs were believed to be a by-product of aberrant splicing or intermediates discarded from intron-lariat debranching \cite{Cocquerelle1993, Qian1992}, going largely unstudied during the formative years of RNA-Seq due to poly-A selection protocols. Recent advancements in bioinformatics methods have facilitated the detection of circRNAs in RNA-Seq datasets, with interest in the field revitalised when Salzman et al. identified circRNAs in cancer and non-cancer cell lines \cite{Salzman2012} resulting in a shift in the paradigm of RNA-Seq protocols enabling detection of circRNAs (total RNA-Seq, ribosomal RNA (rRNA) depleted RNA-Seq). \par
Several characteristics of circRNAs make them ideal candidates as clinical biomarkers for cancers and diseases; lack of 5’ and 3’ ends conferring resistance to RNase degradation granting much higher half-lives than their linear counterparts \cite{Salzman2012, Memczak2013}, tissue- and developmental-stage specific expression, and abundance in blood, serum, plasma and exosomes \cite{Memczak2015, Bahn2015}. Recent evidence has accumulated demonstrating circRNAs viability as a biomarker, with reports of circRNAs exhibiting differential expression in gastric cancer \cite{Li2015Apr}, colorectal cancer \cite{Bachmayr-Heyda2015}, bladder cancer \cite{Zhong2016}, cardiovascular disease \cite{Tan2016} and Alzheimer's disease \cite{Panda2017} amongst many others. circRNAs have also been shown to harbour functionally active and evolutionarily conserved microRNA response elements (MREs) within their mature spliced sequence \cite{Hansen2013, Thomas2014}, suggesting a regulatory role within the competing endogenous RNA (ceRNA) network by titrating the limited pool of endogenous miRNAs available to mRNAs \cite{DENZLER2014}. One such example is circCCDC66, which is hypothesized to play a protective role for the MYC oncogene by binding to miR-33b and miR-99 in colorectal cancer. Knockdown of circCCDC66 displayed reduced MYC expression, whilst administration of miR-33b and miR-99 reversed the knockdown effect, inferring circCCDC66 is capable of modulating mediating MYC expression via the ceRNA network \cite{Hsiao2017May}.
\par
Regarding the quantification, miRNA target prediction and differential expression analysis of circRNAs, several workflows currently exist (Table \ref{tab:my-table}). However, the majority of these workflows are limited by the type of analysis they can perform, often acting as either: 1) a downstream analysis toolkit requiring the user to supply circRNA quantification results, or 2) a workflow offering multiple quantification tools to generate high confidence circRNA calls. Both scenarios necessitate in-house computational expertise to supplement the missing elements of the analysis, resulting in a high barrier to use for both novice and seasoned researchers alike due to the non-standardised input requirements to each workflow.\par

\begin{table}[h!]
\caption{List of current circRNA analysis workflows.}
\begin{tabular}{ccccccc}
\toprule
\textbf{Pipeline} & \textbf{Published} & \textbf{Dependencies} & \textbf{CQ} & \textbf{MP} & \textbf{DE} & \textbf{Installation} \\ \toprule 
CirComPara \cite{circompara} & 2017 & \begin{tabular}[c]{@{}c@{}}Python, R\\ or Docker\end{tabular} & \cmark & \xmark & \begin{tabular}[c]{@{}l@{}}linear\\ RNA\end{tabular} & \begin{tabular}[c]{@{}c@{}}Source code,\\ DockerHub\end{tabular} \\
CirComPara2 \cite{circompara2} & 2021 & \begin{tabular}[c]{@{}l@{}}Python, R\\ or Docker\end{tabular} & \cmark & \xmark & \xmark & \begin{tabular}[c]{@{}c@{}}Source code,\\ DockerHub\end{tabular} \\
DEBKS \cite{DEBKS} & 2021 & Python & \xmark & \xmark & \cmark & \begin{tabular}[c]{@{}c@{}}Source code,\\ Conda, pip\end{tabular} \\
FcircSec \cite{FcircSEC} & 2020 & R & \xmark & \xmark & \xmark & CRAN \\
circRNAProfiler \cite{circRNAprofiler} & 2020 & R & \xmark & \cmark & \cmark & Bioconductor \\
circtools \cite{circtools} & 2019 & Python, R & \cmark & \xmark & \cmark & Conda, Pypi \\
circMeta \cite{Chen2020Jan} & 2020 & R & \xmark & \xmark & \cmark & \begin{tabular}[c]{@{}c@{}}Source code\\ via devtools\end{tabular} \\
circRNAwrap \cite{circRNAwrap} & 2019 & Perl, Python, R & \cmark & \xmark & \cmark & \begin{tabular}[c]{@{}c@{}}Manual\\ Installation\end{tabular} \\
Ularcirc \cite{Ularcirc} & 2019 & R, R Shiny & \xmark & \cmark & \xmark & Bioconductor \\
\textbf{nf-core/circrna} & \textbf{2022} & \textbf{\begin{tabular}[c]{@{}c@{}}nextflow, java,\\ Docker/Apptainer\end{tabular}} & \textbf{\cmark} & \textbf{\cmark} & \textbf{\cmark} & \textbf{DockerHub} \\ \toprule
\multicolumn{7}{l}{\begin{tabular}[c]{@{}l@{}}\textbf{\textit{CQ}} circRNA quantification\\ \textbf{\textit{MP}} miRNA prediction\\ \textbf{\textit{DE}} differential expression\end{tabular}}
\end{tabular}
\label{tab:my-table}
\end{table}

To address these shortcomings, we present nf-core/circrna, a workflow for the quantification, miRNA target prediction and differential expression analysis of circRNAs in RNA-Seq data developed using nextflow \cite{DiTommaso2017Apr}. The pipeline has been developed under the specifications of the nf-core framework \cite{Ewels2020Mar}, ensuring best practices in pipeline development and maintenance. 

\section*{Implementation}
\subsection*{\textbf{nf-core/circrna architecture}}
The workflow is modularised via processes, whereby each process represents an individual task within the workflow. Nextflow automatically distributes each process in an isolated environment where task execution is performed within a container, ensuring efficient utilization and scaling of computing resources user-defined via configuration files. By utilising containers, the arduous task of installing and satisfying software dependencies has been completely removed, thereby facilitating rapid `out of the box' deployment and ensuring reproducible analyses through pinned software versions.
\par
nf-core/circrna is configured to run on local machines, high performance compute clusters and cloud infrastructures alike, facilitating most deployment scenarios via customisable configuration profiles. Profiles for popular platforms (AWS, Azure, Google Cloud) and multiple research institutions can be found at (\url{https://github.com/nf-core/configs}). In the event ones institution is not represented, users are encouraged to fork the nf-core/circrna repository and make the appropriate changes to the configuration profile, reflecting the task scheduler and min/max computational resources available before attempting to deploy the workflow.
\par
Comprehensive documentation including instructions to perform a test dataset analysis can be found at \url{https://nf-co.re/circrna}. Furthermore, the nf-core community harbours an active Slack channel where users can contact the authors directly with queries regarding the workflow, or open an issue on the workflows github page. Three dependencies are required to run nf-core/circrna: Java ($>$=8), the latest version of nextflow \cite{DiTommaso2017Apr} and a software container client such as Docker \cite{Docker} or Apptainer (formerly Singularity) \cite{Singularity}.

\subsection*{\textbf{nf-core/circrna functionality}}
The functionality of nf-core/circrna can be summarized by three core analysis modules: 1) circRNA discovery and quantification 2) miRNA target prediction and 3) differential expression analysis (Figure \ref{fig:1}). The circRNA discovery module is required, however users may choose to include or omit the miRNA prediction or differential expression modules to suit their analysis needs. Pre-processing steps are optional, depending on the format and quality of the input data supplied to the workflow.
\par
The pipeline accepts paired-end RNA-Seq data in the form of FASTQ files or BAM files as inputs to the workflow. In the event BAM files are supplied, the workflow will automatically revert the BAM files to paired end FASTQ files for back-splice junction detection. Summary statistics of read quality, sequencing artefacts and adapter contamination are generated using FastQC \cite{fastqc}, with an optional parameter available to perform adapter removal and read filtering using BBDUK \cite{bbduk}, allowing the user to pass raw or trimmed reads to the circRNA discovery module. 
\par
Reference annotation files (FASTA, GTF) are automatically downloaded via the Illumina iGenomes database, with 18 species currently supported by the workflow in ENSEMBL, UCSC and NCBI annotations where available. In addition to providing correctly formatted reference files to the workflow, the iGenomes database hosts genome index files compatible with multiple aligners, significantly reducing the computational cost of analyses. In the event the user is working with a non-model organism, reference files can be supplied to the workflow manually.

\begin{figure}[h!]
    \includegraphics[width=12.25cm]{figure1.png}
    \vspace{-5mm}
    \caption{Schematic overview of nf-core/circrna analysis modules.}
    \label{fig:1}
\end{figure}

\subsection*{\textbf{circRNA Discovery}}
nf-core/circrna currently offers seven circRNA quantification tools: CircExplorer2 \cite{circexplorer2}, circRNA\_finder \cite{circrna-finder}, DCC \cite{DCC}, CIRIquant \cite{CIRIquant}, MapSplice \cite{MapSplice}, find\_circ \cite{Memczak2013} and Segemehl \cite{segemehl}. As both CIRCexplorer2 and CIRIquant can parse the outputs from multiple aligners, we define CIRCexplorer2 as utilising STAR \cite{STAR} 2-pass mode prior to parsing, whilst CIRIquant utlises its canonical BWA \cite{BWA} - CIRI2 \cite{CIRI} workflow. The user may select one, all or any combination of circRNA tools available in the workflow by providing a list to the `\texttt{-{}-}tool' parameter. In the scenario where more than one quantification tool has been selected, the parameter `\texttt{-{}-}tool\_filter' accepts as input a positive integer \textit{n} to perform consensus based filtering on called circRNAs, requiring circRNAs to be called by at least \textit{n} quantification tools during the count matrix generation step. Users can also apply one of two aggregation functions to circRNAs called by multiple tools using the `\texttt{-{}-}duplicates\_fun' parameter, which accepts `max' or `mean' returning the highest count value or averages, respectively. To reduce spurious circRNA candidates, the parameter `\texttt{-{}-}bsj\_filter' filters circRNAs based on the number of reads spanning the back-splice junction site during parsing of the raw quantification tool outputs.
\par 
Candidates that satisfy the filtering thresholds are passed to an in-house reference-guided circRNA annotation script detailing circRNA type, parent gene, underlying exons and mature spliced length, producing an extended BED-12 output file. The in-house script considers the four following cases for circular candidates: \textit{Circular candidates start and end positions overlap exon boundaries}: If circular candidates start and end positions perfectly overlap exon boundaries, the
candidate is marked as an exonic circRNA, or simply ‘circRNA’. \textit{Circular candidates start and end positions imperfectly overlap exon boundaries}: If the start and end positions are suitably within range ($<$ 200bp) of an exon boundary, the candidate is assumed to be a circRNA. However, if one of the start or end position reside outside of the range of an exon boundary ($>$ 200bp), the circular candidate is assumed to be an exon-intron circRNA (EI-circRNA). Users can toggle the 200bp distance via the `\texttt{-{}-exon\_boundary}' parameter. \textit{Circular candidate resides within a gene, but does not overlap any exon boundaries}: These circular candidates are marked as intronic circular RNAs (ciRNAs). \textit{Circular candidates do not overlap any features in the reference GTF file}: Such candidates are assumed to be intergenic circRNAs and will have an `NA' description for underlying transcripts and parent gene in the output file. Due to the scripts reliance on reference annotations, the workflow is suitable only for analyses of species for which a reference GTF file exists.
\par
Finally, utilising the information held in the BED12 file, the circRNA\_discovery module outputs the mature spliced sequence of each called circRNA in FASTA format and generates a circRNA count matrix for downstream statistical testing, which are used as inputs for the `miRNA\_prediction' and `differential\_expression' modules, respectively. To access the raw output files generated by each quantification tool, users can toggle the `\texttt{-{}-save\_intermediates} boolean parameter to copy files from the work directory to the specified results directory.
\par

\subsection*{\textbf{miRNA prediction}}
To elucidate the regulatory role of circRNAs, nf-core/circrna offers a miRNA prediction module to identify miRNA response element (MRE) sites within the mature spliced sequence and back-splice junction site of detected circRNAs using both miRanda \cite{miRanda} and TargetScan \cite{Targetscan} prediction algorithms. To reduce the number of spurious calls, target miRNAs must be called by both target prediction algorithms. The predictions are output in a results file per sample, detailing circRNA-miRNA node/edge membership complete with miRanda scores, minimum free energy (Kcal/mol) and site type (6mer, 7mer-m8, 7mer-A1, 8mer). Should the user wish to apply strict post-hoc filtering, we suggest users remove 6mers and miRNAs with a minimum free energy score of $>=$-20.00Kcal/Mol \cite{mfe-filter}.

\subsection*{\textbf{Differential expression analysis}}
nf-core/circrna offers a differential expression module to detect differentially expressed circRNAs and model changes in circRNA expression relative to its host gene guided by the phenotype.csv file provided by the user. It is important to note that pre-filtered datasets with linear RNAs removed are not suitable for the differential expression module. \par Firstly, linear RNAs are quantified using Hisat2 \cite{hisat2} and StringTie \cite{StringTie} to determine both host gene expression and library size factors, which are utilised to normalize the circRNA count matrix. Following normalization, quality control plots are generated to assess sample heterogeneity: analysis of principal components via PCA plots and hierarchical clustering of samples via heatmaps and dendograms. The workflow automates differential expression analysis for all possible combinations of the levels provided under the column `condition', whilst controlling for additional covariates added to the phenotype design file using DESeq2 \cite{DESeq2}. For each contrast, the workflow reports both up-regulated and down-regulated circRNAs, plots of global expression patterns via heatmaps, volcano and MA plots, and the distribution of adjusted and non-adjusted p-values. Furthermore, for each differentially expressed circRNA returned by the contrast, boxplots are generated displaying the normalized expression between experimental conditions. To compliment the circRNA results, the workflow concurrently performs linear RNA-Seq differential expression analysis. \par
The CircTest \cite{Jakobi2019Jul} R package employs a beta-binomial model to test differences between circRNA and host gene expression to identify cases of enrichment or depletion of circRNA relative to the linear transcripts from the same host genes. The module outputs PDF plots of circRNA-host gene line plots, ratio plots and a table of the statistical results for users.

\section*{Results and discussion}
\subsection*{\textbf{nf-core/circrna demonstration full dataset}}
To showcase the utility of nf-core/circrna in a single execution, we recapitulated a previously published study investigating the effect of mutant RNA binding homolog FUST-1 on circRNA biogenesis in total RNA seq samples of \textit{C. elegans} (PRJNA742881) \cite{Cao2021Nov} Table \ref{tab:celdataset}. \par

\begin{table}[h!]
\caption{Run information using PRJNA742881 dataset}
\begin{tabular}{cccccc}
\toprule
\textbf{\begin{tabular}[c]{@{}c@{}}Project \\ Accession\end{tabular}} & \textbf{\begin{tabular}[c]{@{}c@{}}Genome \\ Build\end{tabular}} & \textbf{\begin{tabular}[c]{@{}c@{}}Run \\ Time\end{tabular}} & \textbf{\begin{tabular}[c]{@{}c@{}}CPU \\ Hours\end{tabular}} & \textbf{\begin{tabular}[c]{@{}c@{}}Sample\\ ID\end{tabular}} & \textbf{\begin{tabular}[c]{@{}c@{}}Number of \\ reads\end{tabular}} \\ \toprule
\multirow{6}{*}{PRJNA742881} & \multirow{6}{*}{WBcel235} & \multirow{6}{*}{18h 21m 20s} & \multirow{6}{*}{1566} & N2\_1 & 72651039 \\
    &   &   &   & N2\_2  & 92546415  \\
    &   &   &   & N2\_3  & 82870805   \\
    &   &   &   & fust1\_1 & 70194536 \\
    &   &   &   & fust1\_2 & 89945035 \\
    &   &   &   & fust1\_3 & 86470345  \\
\midrule
\end{tabular}
\label{tab:celdataset}
\end{table}

Setting up the analysis required minimal effort -- raw sequencing data was retrieved using nf-core/fetchngs \cite{fetchngs} by providing the sequencing archive project accession number, a sample CSV file was constructed specifying the sample ID's and sample paths, a phenotype CSV file was designed detailing sample metadata for DESeq2 and both the `WBcel235' and `cel' strings were passed to the workflow to automatically download ENSEMBL reference annotation and miRNA database files, respectively. The analysis was run on a local SLURM HPC cluster with maximum ncpu and cpu memory limits set to 8 and 60 GB, representing a moderate level of resources feasible for most research labs. \par

\begin{figure}[h!]
    \centering
    \includegraphics[width=12.25cm]{figure2.png}
    \vspace{-10mm}
    \caption{Comparison of filtered circRNAs \textbf{(A)} circRNAs called by each quantification tool in this study. \textbf{(B)} common circRNAs in this study vs. PRJNA742881. Filtered circRNAs were required to have at least 15 BSJ reads called by CIRIquant. \textbf{(C)} Genomic distribution of 4435 common filtered circRNAs (outer layer), 2420 additional circRNAs discovered by nf-core/circrna (middle layer) and 521 circRNAs unique to PRJNA742881 (inner layer).}
    \label{fig:2}
\end{figure}

\begin{figure}[h!]
    \centering
    \includegraphics[width=12.25cm]{figure3.png}
    \vspace{-10mm}
    \caption{\textbf{(A)} Heatmap of differentially expressed circRNAs \textbf{(B)} Principal component bi-plot of PRJNA742881. \textbf{(C)} Volcano plot of differentially expressed circRNAs.}
    \label{fig:3}
\end{figure}

Based on the results of the simulated dataset (Table \ref{tab:metrics}), CIRCexplorer2, CIRIquant, find\_circ and MapSplice were chosen for circRNA quantification due to their high F1 scores ($>$ 91\%). The original analysis by Cao et. al used CIRCexplorer2, DCC and CIRI2. The justification for not using CIRI2 and DCC are as follows: CIRIquant is the successor to CIRI2, acting as a wrapper script around CIRI2 and as such, CIRI2 remains the underlying circRNA detection software. DCC requires an extremely high amount of RAM (analysis failed with 120GB RAM allocated to DCC processes), most likely caused by the exceptionally high coverage of circRNAs in the RNase R Treatment, Polyadenylation, and Poly(A)$+$ RNA Depleted (RPAD) dataset. Due to it's computational inefficiency and high false positive rate in Table \ref{tab:metrics}, we reasoned that the analysis would not suffer due to its omission. \par
In less than 19 hours, a single execution of nf-core/circrna produced a comprehensive profile of circRNAs in N2 and fust1 \textit{C. elegans} samples, recovering 4,435 filtered circRNAs detected by Cao et al. and an additional 2,420 circRNAs that passed stringent filtering parameters (Figure \ref{fig:2}). A selection of outputs from the `differential\_expression' module are given in Figure \ref{fig:3}, demonstrating the workflow's ability to generate interpretable results for end users. 

\subsection*{\textbf{nf-core/circrna performance evaluation}}
\subsubsection*{\textbf{Computational cost}}
To assess the performance and resource consumption of nf-core/circrna, the run-time and memory (RAM) usage during the analysis of PRJNA742881 was recorded for each process (Figure \ref{fig:4}) with trivial tasks such as file reformatting are omitted from the figure. Per-process usage statistics revealed the most memory intensive processes were the alignment steps for both find\_circ and CIRIquant (38.26 GB, 37.37GB). All other processes in the workflow required minimal resources ($<$ 10 GB). With respect to process run-times, the alignment steps for MapSplice and CIRIquant averaged 209 and 139 minutes, respectively. The customised annotation script averaged a 131 minute run time, however this process is sensitive to the number of circRNA candidates provided to the process, which can be refined using the `-{}-bsj\_reads' parameter. Overall, the performance showcases the efficiency of nf-core/circrna in appropriating tasks in a timely and memory efficient manner, thus making the pipeline deployable on most computing environments. 

\begin{figure}[h!]
    \includegraphics[width=12.25cm]{figure4.png}
    \vspace{-10mm}
    \caption{nf-core/circrna performance results (PRJNA742881). \textbf{A}. Time elapsed per-process in minutes. \textbf{B.} RAM memory consumed per-process in gigabytes.}
    \label{fig:4}
\end{figure}

\subsubsection*{\textbf{Quantification tool performance}}
To assess the performance of nf-core/circrna and it's constituent quantification tools, the workflow was run on a simulated dataset to calculate precision, sensitivity and F1 score performance metrics. The simulated circRNA dataset comprising 1,450,328 synthetic paired-end reads of 101bp in length and 350bp insert size was derived using an overhauled version of CIRIsimulator.pl \cite{CIRI} designed by Zeng et al. \cite{Zeng2017Jun}. This re-worked version of the tool can accept as input circRNA database files from CircBase \cite{circbase}, allowing users to recapitulate \textit{bona fide} circRNAs based on verified back-splice junction coordinates as opposed to the generation of synthetic circRNA reads via the joining of 2 non-collinear exons at random. \par 
7,445 glioblastoma circRNAs were downloaded from circRNADb \cite{circrna_db} and reformatted to conform to CircBase specifications before being supplied as inputs to CIRIsimulator.pl. A total of 7,267 glioblastoma circRNAs were generated by the tool for use in the analysis. As there are no true negatives in the simulated dataset, and all non-true positives are viewed as a false positive, the fraction of relevant instances retrieved by a tool is given by precision, the fraction of relevant instances retrieved relative to the number available instances is given by sensitivity, and the F1 score represents the harmonic mean of precision and sensitivity. In the event a circRNA was called on both strands, we took the decision to  \par
The results of the simulated analysis are given in Table \ref{tab:metrics}. It is important to note sensible filtering parameters were applied to each set of quantification tool results - called circRNAs were required to have at least 2 reads spanning the back-splice junction site. In addition to assessing the performance of each individual quantification tool, circRNAs that had been called by at least two quantification tools via the `\texttt{-{}-}tool\_filter' parameter were extracted from each set of individual results. The resulting set produced scores of 99.83\%, 96.12\% and 97.94\% for precision, sensitivity and F1, respectively, demonstrating the efficacy of a consensus based approach to circRNA quantification. 

\begin{table}[h!]
\caption{Performance metrics of nf-core/circrna quantification tools on the simulated glioblastoma dataset.}
\begin{tabular}{cccccccc}
\toprule
\textbf{Tool}   & \textbf{Detected} & \textbf{\begin{tabular}[c]{@{}c@{}}True \\ Positives\end{tabular}}  & \textbf{Precision} & \textbf{Sensitivity} & \textbf{F1} & \textbf{\begin{tabular}[c]{@{}c@{}}Run \\ Time\end{tabular}}  & \textbf{Memory}  \\ 
\midrule
CIRCexplorer2   & 6270              & 6259        & 99.82\%    & 86.13\%    & 92.47\%     & 7m          & 20.GB         \\
CIRIquant       & 6668              & 6663        & 99.93\%    & 91.69\%    & 95.63\%     & 13m         & 65.GB         \\
circRNA\_finder & 6171              & 5997        & 97.18\%    & 82.52\%    & 89.25\%     & 7m          & 20.GB         \\
DCC             & 8397              & 6344        & 75.55\%    & 87.30\%    & 81.00\%     & 34m         & 21.GB         \\
find\_circ      & 6217              & 6202        & 99.76\%    & 85.34\%    & 91.99\%     & 6m          & 3.GB          \\
MapSplice       & 6155              & 6153        & 99.97\%    & 84.67\%    & 91.69\%     & 9m          & 6.GB          \\
Segemehl        & 4655              & 4558        & 97.92\%    & 62.72\%    & 76.46\%     & 6m          & 50.GB         \\
\midrule
\end{tabular}
\label{tab:metrics}
\end{table}

Another useful metric for users to consider when selecting quantification tools for nf-core/circrna is the proportion of common candidates called by each quantification tool. Previously described by Zeng et al. \cite{Zeng2017Jun}, the proportion of common circRNAs shared between two quantification tools $i$ and $j$ can be represented as $C(i,j)$, with the total number of candidates detected by each tool is $N_i$ and $N_j$, respectively. Thus, for tool $i$, the proportion of common candidates is $P(i,j) = C(i,j)/N_i$ and for tool $j$ the proportion of common candidates is $P(j,i) = C(j,i)/N_j$. Pairwise comparisons for all tools in the simulated dataset were generated and represented using a heatmap (Figure \ref{fig:5}). The heatmap can be read in two directions, permitting the bidirectional query of overlapping sets between two quantification tools. When focusing on a specific column, each cell reflects the proportion of common candidates detected by the column tool that were also detected by the corresponding row tool. Conversely, for each row, the cells represent the proportion of common candidates detected by the corresponding column tool that were also recovered by the row tool. \par
Segemehl detected 4,655 circRNAs with 97.92\% precision in the filtered simulated dataset however, it was unable to recover a high proportion of circRNAs detected by other tools, indicating a conservative approach to calling circRNAs. In contrast, DCC detected 8,397 circRNAs, the highest amount detected by any tool in the simulated dataset. DCC produced a high proportion of circRNAs that were not covered by other tools, suggesting a high rate of false positives - confirmed by its 75.55\% precision score. When using the proportion of common candidates to inform quantification tool choices, users should take caution when using DCC by applying sensible filtering parameters to reduce the number of false positives, and supplement Segemehls conservative results with other quantification tools. Encouragingly, CIRCexplorer2, CIRIquant, find\_circ and MapSplice results were all found to have high agreement - a high proportion of circRNA candidates detected by these methods were frequently detected by one another - and high precision scores, indicating robust methods for detecting circRNAs.

\begin{figure}[h!]
    \centering
    \includegraphics[width=12.25cm]{figure5.png}
    \vspace{-10mm}
    \caption{The proportion of common circRNA candidates called by each quantification tool in the simulated glioblastoma dataset. }
    \label{fig:5}
\end{figure}

\subsubsection*{\textbf{Optimal parameter selection}}
To further advise users on optimal parameter selections for the circrna\_discovery module, we performed an analysis on every possible combination of tools using the simulated dataset coupled with iterative filtering using the `-{}-tool\_filter' parameter. Our goal was to identify parameters which achieve high performance at the lowest computational cost (i.e requiring the fewest amount of quantification tools in a run).  Briefly, a combinatorics approach was taken whereby each set of quantification tool results were used as inputs to generate all possible combinations: $ _{n}C_{k} = \frac{n!}{(n-k)!}$ where $n = 6$ and $k = 1 \dots 6$, resulting in 126 unique sets after the removal of duplicates. In Figure \ref{fig:6}A, the x-axis depicts the number of quantification tools included in the combinatoric sets, whilst the y-axis displays the corresponding mean value for precision, sensitivty and F1 score. Each colored line represents the value selected for the parameter `-{}-tool\_filter', requiring circRNAs in sets to be called by at least \textit{n} tools. \par

\begin{figure}[h!]
    \centering
    \includegraphics[width=12.25cm]{figure6.png}
    \vspace{-10mm}
    \caption{(\textbf{A}) Performance metrics for each combinatorics set of results demonstrating the effect of `-{}-tool\_filter' and number of quantification tools included. (\textbf{B}) T-test of all sets including DCC vs. those without DCC for precision, sensitivity and F1 score. (\textbf{C}) Principal component analysis of the sets, displaying the loadings for precision, sensitivity and F1 score.}
    \label{fig:6}
\end{figure}

We observed an inflection point in the lineplot for F1 scores when three quantification tools were selected and `-{}-tool\_filter' was set to two. Increasing the number of quantification tools resulted in minimal performance gains, whilst applying stricter filtering thresholds via `-{}-tool\_filter' greatly reduced sensitivity. The top ten performing combinations of set size three and `-{}-tool\_filter 2' ranked by F1 score are given in Table \ref{table:topcombinations}. Full performance metric results for all iterations of `-{}-tool\_filter' are available in Additional file 1.
\par 

\begin{table}[ht!]
\caption{Top 10 performing quantification tool combinations at the lowest computational cost in the simulated dataset, ranked by F1 score.}
\centering
\begin{tabular}{lccc}
  \toprule
\textbf{Combination} & \textbf{Precision} & \textbf{Sensitivity} & \textbf{F1} \\ 
  \midrule
CIRCexplorer2,CIRIquant,Segemehl & 98.47\% & 97.21\% & 97.83\% \\ 
  CIRIquant,MapSplice,Segemehl & 98.55\% & 96.99\% & 97.76\% \\ 
  CIRCexplorer2,CIRIquant,find\_circ & 99.59\% & 95.89\% & 97.70\% \\ 
  CIRCexplorer2,CIRIquant,MapSplice & 99.74\% & 95.54\% & 97.60\% \\ 
  CIRIquant,find\_circ,MapSplice & 99.70\% & 95.57\% & 97.59\% \\ 
  CIRCexplorer2,find\_circ,Segemehl & 98.33\% & 96.57\% & 97.45\% \\ 
  CIRCexplorer2,MapSplice,Segemehl & 98.50\% & 96.41\% & 97.44\% \\ 
  CIRIquant,find\_circ,Segemehl & 98.37\% & 96.46\% & 97.41\% \\ 
  CIRCexplorer2,find\_circ,MapSplice & 99.61\% & 95.28\% & 97.40\% \\ 
  CIRCexplorer2,CIRIquant,circRNA\_finder & 97.51\% & 97.18\% & 97.35\% \\ 
   \midrule
\end{tabular}
\label{table:topcombinations}
\end{table}

Furthermore, we were interested in identifying quantification tools that negatively impacted the performance in each of the sets. Clustering analysis of the matrix containing precision, sensitivity and F1 score for all 126 unique sets revealed sets containing results from the tool DCC negatively impacted performance metrics. This was formally tested using a t-test (Figure \ref{fig:6}B) and depicted using principal component analysis whereby the first principal component clearly displays sets containing DCC and negatively correlated with the loadings for precision and F1 score, respectively. DCC is the only quantification tool that utilises STAR 2-pass mode, greatly increasing the sensitivity around splice junction sites at the cost of false positives \cite{Engstrom2013}. Users should take this information into consideration when selecting quantification tools for their own analysis and apply sensible filtering parameters to reduce spurious calls. 


\section*{Conclusion}
nf-core/circrna is the first portable workflow capable of performing the quantification, miRNA target prediction and differential expression analysis of circRNAs in a single execution. Its ease of use greatly reduces the barrier to entry for users seeking to characterise the role of circRNAs in the ceRNA network. We look forward to the workflows release and subsequent engagement with the research community - as the field of circRNA analyses advances, we will effort to incorporate user feedback, suggestions and requests to maintain the vitality and contemporary status of the workflow. 


\section*{Availability and Requirements}
\vspace{0.25cm}
\begin{addmargin}[2.5em]{2.5em}
Project name: nf-core/circrna \\
Project home page: \href{https://nf-co.re/circrna}{https://nf-co.re/circrna} \\
Project github page: \href{https://github.com/nf-core/circrna}{https://github.com/nf-core/circrna} \\
Operating system(s): Platform independent \\
Programming language: Nextflow DSL2\\
Other requirements: Java ($>$=8), nextflow v22.10.4, Docker/Apptainer
License: MIT License \\
Restrictions for academic use: none 
\end{addmargin}

\section*{Supplementary Information}
\textbf{Additional file 1}: \textit{supplementary\_file\_1.xlsx} Data used to produce Figure \ref{fig:6} provided in four xlsx sheets.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Backmatter begins here                   %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{backmatter}

\section*{Abbreviations}
\textbf{\textit{circRNA}} circular RNA \\ \textbf{\textit{ncRNA}} non-coding RNA \\ \textbf{\textit{ceRNA}} competing endogenous RNA \\ \textbf{\textit{rRNA}} ribosomal RNA \\ \textbf{\textit{CQ}} circRNA quantification\\ \textbf{\textit{MP}} miRNA prediction\\ \textbf{\textit{DE}} differential expression\\ \textbf{\textit{miRNA}} microRNA \\ \textbf{\textit{MRE}} microRNA response element \\ \textbf{\textit{ciRNA}} Intronic circRNA \\ \textbf{\textit{EI-circRNA}} Exon-intron circRNA \\ \textbf{\textit{RPAD}} RNase R Treatment, Polyadenylation, and Poly(A)$+$ RNA Depleted 

\section*{Acknowledgements}%% if any
We would like to thank the members of the nf-core community for  technical feedback during the development of the workflow. We also thank the School of Mathematical and Statistical Sciences for hosting presentations of the workflow during its 12\textsuperscript{th} annual research day meeting.

\section*{Authors' contributions}
B.D. designed and developed nf-core/circrna and drafted the manuscript. P.\`O.B. and S.P.F. supervised the work, gave feedback, and revised the manuscript. All authors performed testing on the pipeline and approved the final manuscript.

\section*{Funding}%% if any
Open access and PhD scholarship funding for B.D. provided by Science Foundation Ireland, grant number 18/CRT/6214. No funding body played any roles in the design of the study and collection, analysis, interpretation of data or in writing the manuscript.

\section*{Availability of data and materials}%% if any
The workflow is freely available at \href{https://github.com/nf-core/circrna}{https://github.com/nf-core/circrna}. The full size dataset used in the study is available at \href{https://www.omicsdi.org/dataset/omics_ena_project/PRJNA742881}{PRJNA742881} and code to reproduce the simulated analysis is available at \href{https://github.com/BarryDigby/circRNA_simu}{https://github.com/BarryDigby/circRNA\_simu}. 

\section*{Declarations}

\section*{Ethics approval and consent to participate}%% if any
Not applicable

\section*{Competing interests}
Not applicable

\section*{Consent for publication}%% if any
Not applicable



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                  The Bibliography                       %%
%%                                                         %%
%%  Bmc_mathpys.bst  will be used to                       %%
%%  create a .BBL file for submission.                     %%
%%  After submission of the .TEX file,                     %%
%%  you will be prompted to submit your .BBL file.         %%
%%                                                         %%
%%                                                         %%
%%  Note that the displayed Bibliography will not          %%
%%  necessarily be rendered by Latex exactly as specified  %%
%%  in the online Instructions for Authors.                %%
%%                                                         %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% if your bibliography is in bibtex format, use those commands:
\bibliographystyle{bmc-mathphys} % Style BST file (bmc-mathphys, vancouver, spbasic).
\bibliography{bmcarticle}      % Bibliography file (usually '*.bib' )
% for author-year bibliography (bmc-mathphys or spbasic)
% a) write to bib file (bmc-mathphys only)
% @settings{label, options="nameyear"}
% b) uncomment next line
%\nocite{label}

% or include bibliography directly:
% \begin{thebibliography}
% \bibitem{b1}
% \end{thebibliography}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Figures                       %%
%%                               %%
%% NB: this is for captions and  %%
%% Titles. All graphics must be  %%
%% submitted separately and NOT  %%
%% included in the Tex document  %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%
%% Do not use \listoffigures as most will included as separate files

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Tables                        %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Additional Files              %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\end{backmatter}
\end{document}
